### PR TITLE
Add basic tests for cloud shell

### PR DESCRIPTION
**Pull request by flacatus**
Signed-off-by: flacatus <flacatus@redhat.com>

### What does this PR do?
This PR contains necessary files to start to use e2e tests for Che-Workspace-operator in Prow CI. The tests includes:
- Setup ginkgo (testing golang framework)
- Deploy plugin registry
- Create roles and routes in OCP cluster
- Deploy Workspace Operator
- Start cloud shell dev workspaces

This tests will be part of PROW CI running in latest stable Openshift and nightly version. More info about PROW: https://ci-operator-configresolver-ui-ci.apps.ci.l2s4.p1.openshiftapps.com/ and  
https://docs.google.com/document/d/1SQ_qlkcplqhe8h6ONXdgBr7YUVbs4oRSj4ISl3gpLW4/edit

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/16667

### CODE REVIEW

**Comment by flacatus** [comment]
cc @rhopp

File: test/e2e/cmd/workspaces_test.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+var _ = ginkgo.SynchronizedAfterSuite(func() {
+	workspaces := workspaces.NewWorkspaceClient()
+
+	if err := workspaces.Kube().CoreV1().Namespaces().Delete(metadata.Namespace.Name, &metav1.DeleteOptions{}); err != nil {
+		_ = fmt.Errorf("Failed to deploy workspace controller %s", err)
+	}
+}, func() {})
+
+func TestHarnessCodeReadyWorkspaces(t *testing.T) {
```
Comment (JPinkney) [inline test/e2e/cmd/workspaces_test.go:79]: This is technically upstream, I don't think we should have references to CRW

Comment (sleshchenko) [reply test/e2e/cmd/workspaces_test.go:79]: I believe it's just copy/paste mistake that is going to be fixed

File: test/e2e/pkg/workspaces/workspaces.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+	deploy, err := w.PodDeployWaitUtil(label)
+	if !deploy {
+		fmt.Println("Che Workspaces Controller not deployed")
+		return err
+	}
+	return nil
+}
+
+func (w *CodeReady) CreatePluginRegistryService() (err error) {
```
Comment (amisevsk) [inline test/e2e/pkg/workspaces/workspaces.go:33]: I don't know how useful a named return is here, it'd be cleaner to not have it. It's also cleaner to not define vars here unless necessary, as it makes it easier to miss an error check.

File: test/e2e/pkg/metadata/metadata.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package metadata
+
+type OperatorNamespace struct {
+	Name string
+}
+
+var Namespace = OperatorNamespace{}
```
[the remark this answers is not in the document]
Comment (flacatus) [reply test/e2e/pkg/metadata/metadata.go:7]: Would be useful to have metadata to get in future data from pods for example...

File: test/e2e/pkg/workspaces/workspaces.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	deserializeService, _ := deserializePluginRegistryService()
```
[the remark this answers is not in the document]
Comment (flacatus) [reply test/e2e/pkg/workspaces/workspaces.go]: I delete this function

File: test/e2e/pkg/tests/cloud_shell_tests.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package tests
+
+import (
+	"fmt"
+	"github.com/che-incubator/che-workspace-operator/test/e2e/pkg/workspaces"
+	"github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
```
[the remark this answers is not in the document]
Comment (flacatus) [reply test/e2e/pkg/tests/cloud_shell_tests.go]: Done.

File: test/e2e/cmd/workspaces_test.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	"fmt"
+	"path/filepath"
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/che-incubator/che-workspace-operator/test/e2e/pkg/metadata"
+	_ "github.com/che-incubator/che-workspace-operator/test/e2e/pkg/tests"
+	workspaces "github.com/che-incubator/che-workspace-operator/test/e2e/pkg/workspaces"
```
[the remark this answers is not in the document]
Comment (flacatus) [reply test/e2e/cmd/workspaces_test.go]: Done

File: test/e2e/pkg/workspaces/client.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+type CodeReady struct {
+	rest *rest.Config
+}
+
+// NewCodeReady creates C, a workspaces used to expose common testing functions.
```
[the remark this answers is not in the document]
Comment (flacatus) [reply test/e2e/pkg/workspaces/client.go]: My mistake

File: test/e2e/pkg/workspaces/oc.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	output, err := cmd.CombinedOutput()
+	fmt.Println(string(output))
+	if err != nil && !strings.Contains(string(output), "AlreadyExists") {
+		fmt.Println(err)
+	}
+	return err
+}
+
+func (w *CodeReady) CustomResourceDefinitions() (err error) {
+	cmd := exec.Command("oc", "apply", "--namespace", metadata.Namespace.Name, "-f", "deploy/crds")
```
[the remark this answers is not in the document]
Comment (flacatus) [reply test/e2e/pkg/workspaces/oc.go]: Right, Changed!

File: test/e2e/cmd/workspaces_test.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+//SynchronizedBeforeSuite blocks is executed before run all test suites
+var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
+	//!TODO: Try to create a specific function to call all <ginkgo suite> configuration.
+	fmt.Println("Starting to setup objects before run ginkgo suite")
+	config.Namespace = "che-workspace-controller"
+
+	workspaces, err := client.NewK8sClient()
+	if err != nil {
+		fmt.Println("Failed to create workspace client")
```
Comment (sleshchenko) [inline test/e2e/cmd/workspaces_test.go:47]: If you don't fail tests here - the code will fail with  nil error at line 51

File: test/e2e/cmd/workspaces_test.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	jUnitOutputFilename  = "junit-workspaces-operator.xml"
+)
+
+//SynchronizedBeforeSuite blocks is executed before run all test suites
+var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
+	//!TODO: Try to create a specific function to call all <ginkgo suite> configuration.
+	fmt.Println("Starting to setup objects before run ginkgo suite")
+	config.Namespace = "che-workspace-controller"
+
+	workspaces, err := client.NewK8sClient()
```
Comment (sleshchenko) [inline test/e2e/cmd/workspaces_test.go:45]: I think it would be a bit better to name it k8sClient

File: test/e2e/cmd/workspaces_test.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	controller := deploy.NewDeployment(workspaces)
+
+	if err := controller.CreateAllOperatorRoles(); err != nil {
+		_ = fmt.Errorf("Failed to create roles in clusters %s", err)
+	}
+
+	if err := controller.CreateOperatorClusterRole(); err != nil {
+		_ = fmt.Errorf("Failed to create roles in clusters %s", err)
```
Comment (sleshchenko) [inline test/e2e/cmd/workspaces_test.go:64]:
```suggestion
		_ = fmt.Errorf("Failed to create cluster roles in clusters %s", err)
```

File: test/e2e/cmd/workspaces_test.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	if err := controller.CustomResourceDefinitions(); err != nil {
+		_ = fmt.Errorf("Failed to add custom resources definitions to cluster %s", err)
```
Comment (sleshchenko) [inline test/e2e/cmd/workspaces_test.go:68]: let's fail in any error, if deployment is failed - the tests are supposed to fail as well.

File: test/e2e/pkg/metadata/metadata.go (hunk 3 above)
Comment (sleshchenko) [inline test/e2e/pkg/metadata/metadata.go:7]: It's unused now

File: test/e2e/cmd/workspaces_test.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+	fmt.Println("Creating ginkgo reporter for Test Harness: Junit and Debug Detail reporter")
+	var r []ginkgo.Reporter
+	r = append(r, reporters.NewJUnitReporter(filepath.Join(testResultsDirectory, jUnitOutputFilename)))
+
+	fmt.Println("Running Workspace Controller e2e tests...")
+	ginkgo.RunSpecsWithDefaultAndCustomReporters(t, "Workspaces Controller Operator Tests", r)
+
+}
+
+func newNamespace() (ns *corev1.Namespace) {
```
Comment (sleshchenko) [inline test/e2e/cmd/workspaces_test.go:102]: I think it makes sense to move it to deployment.

File: test/e2e/pkg/deploy/controller.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+		return err
+	}
+	return err
+}
+
+func (w *Deployment) CreateAllOperatorRoles() (err error) {
+	cmd := exec.Command("oc", "apply", "--namespace", config.Namespace, "-f", "deploy")
+	output, err := cmd.CombinedOutput()
+	fmt.Println(string(output))
+	if err != nil && !strings.Contains(string(output), "AlreadyExists") {
```
Comment (sleshchenko) [inline test/e2e/pkg/deploy/controller.go:45]: If the purpose of AlreadyExists check - is not to fail the tests - then code should be fixed.

File: test/e2e/pkg/client/pod.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+func (w *K8sClient) PodDeployWaitUtil(label string) (deployed bool, err error) {
+	timeout := time.After(15 * time.Minute)
+	tick := time.Tick(1 * time.Second)
+
+	for {
+		select {
+		case <-timeout:
+			return false, errors.New("timed out")
+		case <-tick:
+			desc := w.WaitForPodBySelectorRunning(config.Namespace, label, 180)
```
Comment (sleshchenko) [inline test/e2e/pkg/client/pod.go:36]: what desc mean here?

File: test/e2e/pkg/metadata/metadata.go (hunk 3 above)
Comment (flacatus) [reply test/e2e/pkg/metadata/metadata.go:7]: Will be necessary in the future to collect data in PROW

File: test/e2e/pkg/client/client.go [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+type K8sClient struct {
+	kubeClient *kubernetes.Clientset
+}
+
+// NewCodeReady creates C, a workspaces used to expose common testing functions.
```
Comment (sleshchenko) [inline test/e2e/pkg/client/client.go:25]:
```suggestion
// NewK8sClient creates kubernetes client wrapper with helper functions and direct access to k8s go client
```

File: test/e2e/pkg/client/pod.go [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+		switch pod.Status.Phase {
+		case v1.PodRunning:
+			fmt.Println("Pod started after", age, "seconds")
+			return true, nil
+		case v1.PodFailed, v1.PodSucceeded:
+			return false, nil
+		}
+		return false, nil
+	}
+}
```
Comment (sleshchenko) [inline test/e2e/pkg/client/pod.go:100]:
```suggestion
}

```

**Pull request merged**